src/Lithnet.GoogleApps/CourseStudentRequestFactory.cs

```diff
@@ -0,0 +1,405 @@
+        {
+            try
+            {
+                this.WaitForGate();
+
+                using (PoolItem<ClassroomService> poolService = this.classRoomServicePool.Take(NullValueHandling.Ignore))
+                {
+                    CoursesResource.StudentsResource.DeleteRequest request = poolService.Item.Courses.Students.Delete(courseId, userId);
+                    Trace.WriteLine($"Removing student {userId} from course {courseId}");
+                    try
+                    {
+                        request.ExecuteWithRetry(RetryEvents.Backoff | RetryEvents.Timeout, RetryCount);
+                    }
+                    catch (GoogleApiException e)
+                    {
+                        if (!throwOnMissingStudent && this.IsMissingStudentError(e.HttpStatusCode, e.Message))
+                        {
+                            return;
+                        }
+
+                        throw;
+                    }
+                }
+            }
+            finally
+            {
+                this.ReleaseGate();
+            }
+        }
+
+        public void AddStudents(string courseId, IList<Student> students, bool throwOnExistingStudent)
+        {
+            if (CourseStudentRequestFactory.BatchSize <= 1)
+            {
+                foreach (Student student in students)
+                {
+                    this.AddStudent(courseId, student, throwOnExistingStudent);
+                }
+
+                return;
+            }
+
+            try
+            {
+                this.WaitForGate();
+
+                using (PoolItem<ClassroomService> poolService = this.classRoomServicePool.Take(NullValueHandling.Ignore))
+                {
+                    List<ClientServiceRequest<Student>> requests = new List<ClientServiceRequest<Student>>();
+
+                    foreach (Student student in students)
+                    {
+                        Trace.WriteLine($"Queuing batch student add for {student.UserId} to course {courseId}");
+                        requests.Add(poolService.Item.Courses.Students.Create(student, courseId));
+                    }
+                    this.ProcessBatches<Student>(courseId, !throwOnExistingStudent, false, requests, poolService, (batchHelper) =>
+                    {
+                        return (content, error, i, message) =>
+                        {
+                            int index = batchHelper.BaseCount + i;
+                            this.ProcessStudentResponse(courseId, (students[index] as Student).UserId, batchHelper.IgnoreExistingStudent, batchHelper.IgnoreMissingStudent, error, message, batchHelper.RequestsToRetry, batchHelper.Request, batchHelper.FailedStudents, batchHelper.Failures);
+                        };
+                    });
+                }
+            }
+            finally
+            {
+                this.ReleaseGate();
+            }
+        }
+
+
+        protected class ProcessBatchHelper<T>
```

Comment:
Can we move this into it's own file. Should it be 'protected internal'?

Reply:
Done. Moved to own file CourseStudentRequestFactory.ProcessBatchHelper.cs. Commit coming.

Reply:
Also, class is now protected internal.